report_data.py

```diff
@@ -213,20 +260,46 @@ def process_sales(self, df_sales):
         # - track "fee of sale" in "fees"
         # - track net proceeds in held_forex
         for row_idx, row in df_sales.iterrows():
-            sold_quantity = row.quantity
+            sold_quantity = to_decimal(row.quantity)
             sold_symbol = row.symbol
 
             if sold_quantity < 0:
-                raise ValueError(f"In 'sales' tab, row number {row_idx + 2} for symbol '{sold_symbol}' the quantity is negative")
+                raise ValueError(
+                    f"In 'sales' tab, row number {row_idx + 2} for symbol '{sold_symbol}' the quantity is negative"
+                )
+
+            # continously check whether stock splits are applicable
+            stock_splits_applicable = True
+            while stock_splits_applicable:
+                if len(self.stock_splits[sold_symbol]) == 0:
+                    stock_splits_applicable = False
+                elif row.date <= self.stock_splits[sold_symbol][0].date:
+                    # sell orders before any split don't need to consider stock splits
+                    stock_splits_applicable = False
+                elif self.held_shares[sold_symbol].is_empty():
+                    # will raise an error later
+                    stock_splits_applicable = False
+                elif (
+                    self.stock_splits[sold_symbol][0].date
+                    < self.held_shares[sold_symbol].peek().buy_date
+                ):
+                    # pop splits which are older than any held share anyways
+                    self.stock_splits[sold_symbol].pop(0)
+                else:
+                    # apply splits for all held shares which have been bought
+                    # after the split, until date of sold share is before
```

Comment:
> until date of sold share is before next potential split

I think you need to apply it starting from the buy date up until the end if the buy date is before the split date, not just until the next split.

In the next example I assumed the stock price remains constant.

* Buy date: 1999-12-31: 100 stocks

Split: 2000-01-01: 1:10 (1 stock -> 10 stocks)

* Sell: 2001-01-01: 100 stocks
* Sell: 2002-01-01: 100 stocks
* Remaining: 800 stocks

Split: 2010-01-01: 1:10 (1 stock -> 10 stocks)

* Sell: 2011-01-01: 8000 stocks
* Remaining: 0

In order to get the 8000 stocks, you had to multiply the original 100 stocks twice (at the first split and and at the second split)

Reply:
Writing these things, I actually realized that it would make more sense to rewrite the whole processing logic anyways. Right now, I first process all buys at once. 
Technically, it would make more sense to keep a state (Forex balances, FIFO Queues for shares, etc.) and keep all buy orders / sell orders / dividend payments / stock splits as events which are put into a single ordered list of events. Processing the stock split then is just a lot easier.

Reply:
Right now, I think mainly the comment is wrong. When processing a sale order, you every time check whether stock splits have to be applied first. You iterate over all of the applicable splits. `FIFOQueue.apply_split` applies the split for all stocks in the FIFOQueue whose buy date is before the split date. The comment does not make any sense. But overall, the logic should be sane but ugly.